src/dataflow/mod.rs

```diff
@@ -684,64 +691,134 @@ pub(crate) enum Step {
     },
 }
 
-/// Decode Steps from Python (name, funcs...) tuples.
+/// Decode Steps from Python (name, funcs...) dict.
 ///
 /// Required for pickling.
 impl<'source> FromPyObject<'source> for Step {
     fn extract(obj: &'source PyAny) -> PyResult<Self> {
-        let tuple: &PySequence = obj.downcast()?;
-        if let Ok(("Input", step_id, input_config)) = tuple.extract() {
-            Ok(Self::Input {
-                step_id,
-                input_config,
-            })
-        } else if let Ok(("Map", mapper)) = tuple.extract() {
-            Ok(Self::Map { mapper })
-        } else if let Ok(("FlatMap", mapper)) = tuple.extract() {
-            Ok(Self::FlatMap { mapper })
-        } else if let Ok(("Filter", predicate)) = tuple.extract() {
-            Ok(Self::Filter { predicate })
-        } else if let Ok(("FoldWindow", step_id, clock_config, window_config, builder, folder)) =
-            tuple.extract()
-        {
-            Ok(Self::FoldWindow {
-                step_id,
-                clock_config,
-                window_config,
-                builder,
-                folder,
-            })
-        } else if let Ok(("Inspect", inspector)) = tuple.extract() {
-            Ok(Self::Inspect { inspector })
-        } else if let Ok(("InspectEpoch", inspector)) = tuple.extract() {
-            Ok(Self::InspectEpoch { inspector })
-        } else if let Ok(("Reduce", step_id, reducer, is_complete)) = tuple.extract() {
-            Ok(Self::Reduce {
-                step_id,
-                reducer,
-                is_complete,
-            })
-        } else if let Ok(("ReduceWindow", step_id, clock_config, window_config, reducer)) =
-            tuple.extract()
-        {
-            Ok(Self::ReduceWindow {
-                step_id,
-                clock_config,
-                window_config,
-                reducer,
-            })
-        } else if let Ok(("StatefulMap", step_id, builder, mapper)) = tuple.extract() {
-            Ok(Self::StatefulMap {
-                step_id,
-                builder,
-                mapper,
-            })
-        } else if let Ok(("Capture", output_config)) = tuple.extract() {
-            Ok(Self::Capture { output_config })
-        } else {
-            Err(PyValueError::new_err(format!(
-                "bad python repr when unpickling Step: {obj:?}"
-            )))
+        let dict: &PyDict = obj.downcast()?;
+        let t: &str = dict
+            .get_item("type")
+            .expect("Unable to extract step, missing `type` field.")
+            .extract()?;
```

Comment:
Since this is used a lot here, maybe wrap it in a function:
```rust
fn step_extract<'a, D>(dict: &'a PyDict, key: &str) -> PyResult<D>
where
    D: FromPyObject<'a>,
{
    dict.get_item(key)
        .expect("Unable to extract step, missing `{key}` field.")
        .extract()
}
```

Reply:
Same for all the steps below

Reply:
```rust
let t: &str = step_extract(dict, "type")?;
```

```rust
        match t {
            "Input" => Ok(Self::Input {
                step_id: step_extract(dict, "step_id")?,
                input_config: step_extract(dict, "input_config")?,
            }),
        ...
```